[Data Exfiltration Detection] Update and add anomaly detection jobs and security rules

sodhikirti07 opened the pull request:
<!-- Type of change
Please label this PR with one of the following labels, depending on the scope of your change:
- Bug
- Enhancement
- Breaking change
- Deprecation
-->

## What does this PR do?
- Adds filter to map source IP and destination IP to only contain internal and external addresses respectively.
- Changes the configurations of current anomaly detection jobs.
- Changed the description of detectors in ml module
- Adds two anomaly detection jobs and rules to identify exfiltration to external devices (USB, Airdrop)
- Updates the lookback time of the detection rules i.e. 2*bucket_span of ml job.
- Updates README, changelog, manifest and dashboard.

<!-- Mandatory
Explain here the changes you made on the PR.
-->

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [x] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).

## Author's Checklist

<!-- Recommended
Add a checklist of things that are required to be reviewed in order to have the PR approved
-->
- [ ]

## How to test this PR locally


<!-- Recommended
Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.
-->

## Related issues

<!-- Recommended
Link related issues below. Insert the issue link or reference after the word "Closes" if merging this should automatically close it.

- Closes #123
- Relates #123
- Requires #123
- Supersedes #123
-->
-

## Screenshots

<!-- Optional
Add here screenshots presenting:
- Kibana UI forms presenting configuration options exposed by the integration
- dashboards with collected metrics or logs
-->
Find them in the comment section

sodhikirti07 commented:
### Screenshots

- Overview page

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/integrations/builds/105)
* Commit: 70bbe03c70e0ca316d76c0a84ebf197f5b1329d1

### History
* :green_heart: [Build #104](https://buildkite.com/elastic/integrations/builds/104) succeeded cb33c912315a4b2fe07867f16c07a2c364ff2d59

<!--buildkite-pr-comment-integrations
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/105","number":105,"commit":"70bbe03c70e0ca316d76c0a84ebf197f5b1329d1"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/104","number":104,"commit":"cb33c912315a4b2fe07867f16c07a2c364ff2d59"}],"number":105}
buildkite-pr-comment-->

elasticmachine commented:
## :green_heart: Build Succeeded


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-6577/detail/PR-6577/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-6577/detail/PR-6577/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-6577/detail/PR-6577/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-6577/detail/PR-6577/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://integrations_6577.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/fleet-ci/transactions/view?rangeFrom=2023-06-14T14:01:39.654Z&rangeTo=2023-06-14T14:21:39.654Z&transactionName=BUILD+Ingest-manager%2Fintegrations%2FPR-%7Bnumber%7D&transactionType=job&latencyAggregationType=avg&traceId=082ffdbfb80c3a358aaac2169d5f1e83&transactionId=2747080ea4d1f24b)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2023-06-14T14:11:39.654+0000


* Duration: 13 min 14 sec




#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 1  |
| Skipped      | 0 |
| Total        | 1   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->














## :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.




</p>
</details>

<!--COMMENT_GENERATED_WITH_ID_comment.id-->

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/integrations/builds/108)
* Commit: 9df2bb9f1f20d586ccddd70ab2b0d682e3103243

### History
* :green_heart: [Build #105](https://buildkite.com/elastic/integrations/builds/105) succeeded 70bbe03c70e0ca316d76c0a84ebf197f5b1329d1
* :green_heart: [Build #104](https://buildkite.com/elastic/integrations/builds/104) succeeded cb33c912315a4b2fe07867f16c07a2c364ff2d59

<!--buildkite-pr-comment-integrations
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/108","number":108,"commit":"9df2bb9f1f20d586ccddd70ab2b0d682e3103243"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/105","number":105,"commit":"70bbe03c70e0ca316d76c0a84ebf197f5b1329d1"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/104","number":104,"commit":"cb33c912315a4b2fe07867f16c07a2c364ff2d59"}],"number":108}
buildkite-pr-comment-->

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/integrations/builds/109)
* Commit: 0181dff7d16e6f14cb42bea3460ca1d6127eb600

### History
* :green_heart: [Build #108](https://buildkite.com/elastic/integrations/builds/108) succeeded 9df2bb9f1f20d586ccddd70ab2b0d682e3103243
* :green_heart: [Build #105](https://buildkite.com/elastic/integrations/builds/105) succeeded 70bbe03c70e0ca316d76c0a84ebf197f5b1329d1
* :green_heart: [Build #104](https://buildkite.com/elastic/integrations/builds/104) succeeded cb33c912315a4b2fe07867f16c07a2c364ff2d59

<!--buildkite-pr-comment-integrations
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/109","number":109,"commit":"0181dff7d16e6f14cb42bea3460ca1d6127eb600"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/108","number":108,"commit":"9df2bb9f1f20d586ccddd70ab2b0d682e3103243"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/105","number":105,"commit":"70bbe03c70e0ca316d76c0a84ebf197f5b1329d1"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/104","number":104,"commit":"cb33c912315a4b2fe07867f16c07a2c364ff2d59"}],"number":109}
buildkite-pr-comment-->

susan-shu-c commented:
I feel that I'd avoid creating a lot of false positives for new hosts 🤔 re: `host.name`

2nd question, in the description's second point, `Changes the configurations of current anomaly detection jobs.`, what are the main changes?

Lastly, is it possible to explain how the Airdrop was tested?

peteharverson commented on packages/ded/kibana/ml_module/ded-ml.json:
```diff
@@ -161,7 +168,8 @@
                             "source.ip",
-                            "destination.ip"
+                            "destination.ip",
+                            "destination.geo.city_name"
```
What was the reason for going with `geo.city_name` as an extra influencer, but not using `region_name`?

sodhikirti07 replied:
Added for additional context only, `region_name` is already there as a field.

peteharverson replied:
But `destination.geo.region_name` is not included in the `influencers` for this job. The `over` field name would normally be included as an influencer.

sodhikirti07 commented:
> 2nd question, in the description's second point, Changes the configurations of current anomaly detection jobs., what are the main changes?

@susan-shu-c 
I updated the filter query to include internal source and external destination IPs. The anomaly detection jobs are now partitioned by `host.name` , earlier we were doing population analysis. Lastly, deleted some redundant jobs like anomalies looking at city names, continent names and added them as influencers instead (for info gain). See the current jobs [here](https://github.com/elastic/integrations/blob/main/packages/ded/kibana/ml_module/ded-ml.json)

> Lastly, is it possible to explain how the Airdrop was tested?

Tested by setting up a macOS VM using Parallels desktop and also confirmed with Ricardo Ungureanu.

peteharverson commented on packages/ded/kibana/ml_module/ded-ml.json:
```diff
@@ -30,21 +30,22 @@
                         "detectors": [
                             {
-                                "detector_description": "high_sum(\"source.bytes\") over \"destination.geo.city_name\"",
+                                "detector_description": "high_sum(\"source.bytes\") over \"destination.geo.country_iso_code\"",
```
All the detectors in the jobs are using the default descriptions which are not always the most user-friendly. These are used in various places in the UI - for example in the ML anomalies table. Any thoughts on going with something which is closer to the form of text used in the job descriptions? For example `High bytes sent to an unusual country_iso_code`? I know we often go with the defaults, but just wondering if there is something we could do here?

sodhikirti07 replied:
TBH, I didn't try to touch them as we always followed the default for these jobs. I'm down trying simple text as a descriptor, but IDK, If it's okay to deviate one package's structure from the rest.

ajosh0504 replied:
I'd agree with Pete. We've switched to what he's suggesting for our pre-built ML rules. I'd say let's change this package to begin with, and also update others in later versions.

sodhikirti07 replied:
Changed the description of detectors

![image](https://github.com/elastic/integrations/assets/109447885/637182a2-e401-481c-a03c-103b561fe555)

peteharverson reviewed (commented):
Overall looks like some great updates here. Left a couple of comments on the detector descriptions and the influencers for one of the ML jobs.

ajosh0504 commented on packages/ded/docs/README.md:
```diff
@@ -10,37 +10,37 @@ Then use these detection rules and anomaly detection jobs for data exfiltration
+| Potential Data Exfiltration Activity to an Unusual Region | An anomaly detection job to detect an abnormal volume of bytes being sent to an unusual region name.           |
+| Potential Data Exfiltration Activity to an Unusual IP Address | An anomaly detection job to detect an abnormal volume of bytes being sent to an unusual IP address.            |
+| Potential Data Exfiltration Activity to an Unusual Destination Port | An anomaly detection job to detect an abnormal volume of bytes being sent to an unusual destination port.      |
+ | High bytes of data sent to an external device | An anomaly detection job to detect high bytes of data written to an external device.                           |
```
Please change rule names to Title Case.

ajosh0504 commented on packages/ded/docs/README.md:
```diff
@@ -10,37 +10,37 @@ Then use these detection rules and anomaly detection jobs for data exfiltration
+| Potential Data Exfiltration Activity to an Unusual IP Address | An anomaly detection job to detect an abnormal volume of bytes being sent to an unusual IP address.            |
+| Potential Data Exfiltration Activity to an Unusual Destination Port | An anomaly detection job to detect an abnormal volume of bytes being sent to an unusual destination port.      |
+ | High bytes of data sent to an external device | An anomaly detection job to detect high bytes of data written to an external device.                           |
+| High bytes of data sent to an external device via Airdrop | An anomaly detection job to detect high bytes of data written to an external device via Airdrop. |
```
Please change rule names to Title Case.

ajosh0504 commented on packages/ded/docs/README.md:
```diff
@@ -10,37 +10,37 @@ Then use these detection rules and anomaly detection jobs for data exfiltration
+| Potential Data Exfiltration Activity to an Unusual Destination Port | An anomaly detection job to detect an abnormal volume of bytes being sent to an unusual destination port.      |
+ | High bytes of data sent to an external device | An anomaly detection job to detect high bytes of data written to an external device.                           |
+| High bytes of data sent to an external device via Airdrop | An anomaly detection job to detect high bytes of data written to an external device via Airdrop. |
+ | Rare process writing data to an external device | An anomaly detection job to detect a rare process writing data to an external device.                          |
```
Please change rule names to Title Case.

ajosh0504 commented on packages/ded/kibana/ml_module/ded-ml.json:
```diff
@@ -138,21 +144,22 @@
             },
             {
-                "id": "high-sent-bytes-destination-geo-country_name",
+                "id": "high-sent-bytes-destination-region_name",
```
Any reason for retaining the region name job, in addition to ISO code?

sodhikirti07 replied:
They both have different set of granularities, so thought would be nice to see anomalies at both levels.

ajosh0504 commented on packages/ded/kibana/ml_module/ded-ml.json:
```diff
@@ -210,30 +219,31 @@
                         "data_exfiltration"
                     ],
-                    "description": "A machine learning job to detect data exfiltration to an unusual destination port.",
+                    "description": "Detects data exfiltration activity by identifying a writing event started by a rare process to an external device.",
```
```suggestion
                    "description": "Detects data exfiltration activity by identifying a file write started by a rare process to an external device.",
```

ajosh0504 commented on packages/ded/kibana/security_rule/798c1290-6411-486e-bc85-7ae9fe826aaf.json:
```diff
@@ -0,0 +1,27 @@
+        "interval": "15m",
+        "license": "Elastic License",
+        "machine_learning_job_id": "high-bytes-written-to-external-device",
+        "name": "High bytes of data sent to an external device",
```
```suggestion
        "name": "Spike in Bytes Sent to an External Device",
```

ajosh0504 commented on packages/ded/kibana/security_rule/7a342a6b-58df-4832-919d-d4bd681e3138.json:
```diff
@@ -0,0 +1,27 @@
+        "interval": "15m",
+        "license": "Elastic License",
+        "machine_learning_job_id": "rare-process-writing-to-external-device",
+        "name": "Rare process writing data to an external device",
```
```suggestion
        "name": "Unusual Process Writing Data to an External Device",
```

ajosh0504 commented on packages/ded/kibana/security_rule/da5cd97e-7f3c-40dc-9183-7ef1270e8817.json:
```diff
@@ -0,0 +1,27 @@
+        "interval": "15m",
+        "license": "Elastic License",
+        "machine_learning_job_id": "high-bytes-written-to-external-device-airdrop",
+        "name": "High bytes of data sent to an external device via Airdrop",
```
```suggestion
        "name": "Spike in Bytes Sent to an External Device via Airdrop",
```

ajosh0504 reviewed (approved):
Love the new detections! Made some minor comments.

peteharverson reviewed (approved):
Changes for the ML jobs LGTM

sodhikirti07 commented:
Made the following changes based on my discussion with Dian:
- Changed the ml-module query to work with either file or network events.
- Changed the anomaly detection jobs to their original state, i.e., removed partition by `host.name`. The `host.name` field is not compatible with all the data sources for network logs. Identifiers such as `source.ip` are the most applicable here but creates a hard limit on the model due to their high cardinalities.

Haven't changed the partition by field in the exfiltration to external device jobs. Let me know the consensus of folks in using `user.name` or `host.name` as the partitioning field.

cc: @ajosh0504 @susan-shu-c @dainperkins

susan-shu-c commented:
To document our discussion, I'll defer to Dain for now on using `user`, though agree that there are places it's not so clear-cut.

ajosh0504 commented on packages/ded/kibana/ml_module/ded-ml.json:
```diff
@@ -2,19 +2,27 @@
                         }
                     },
+                    {
+                        "bool":{
```
Do you need this bool? Also do you have to check for existence of source bytes and destination fields? Are they not present in all network events?

ajosh0504 commented on packages/ded/kibana/security_rule/5576098d-650d-4b0f-a3a1-48b5b0ce4f15.json:
```diff
@@ -20,7 +20,7 @@
         ],
         "type": "machine_learning",
-        "version": 3
+        "version": 5
```
Should be 4?

sodhikirti07 replied:
Ah! yes, thanks for catching.

ajosh0504 commented on packages/ded/kibana/security_rule/8d4e2e5f-982b-4e8d-b765-a10d648efc33.json:
```diff
@@ -20,7 +20,7 @@
         ],
         "type": "machine_learning",
-        "version": 3
+        "version": 5
```
Should be 4?

ajosh0504 commented on packages/ded/kibana/security_rule/b9d01e71-ac5d-4203-82ed-c1246d6c72c8.json:
```diff
@@ -20,7 +20,7 @@
         ],
         "type": "machine_learning",
-        "version": 3
+        "version": 5
```
Should be 4?

sodhikirti07 commented on packages/ded/kibana/ml_module/ded-ml.json:
```diff
@@ -2,19 +2,27 @@
+                            "event.category": "file"
                         }
                     },
+                    {
```
Done!

sodhikirti07 commented:
Based on my discussion with @susan-shu-c and @ajosh0504, I am keeping the current `partition_by` configurations in place i.e., partitioning the jobs by `host.name`.

elasticmachine commented:
Package ded - 1.0.3 containing this change is available at https://epr.elastic.co/search?package=ded

dainperkins commented:
2 things - 
1) USB exfil - from a security and behavior modeling standpoint,  in any environment with shared workstations this will work much better if partitioned by user, as hosts will invariably be less aligned to a specific behavior profile, and it seemed like most file events will have a user involved

2) if the partitioning by host.name for network exfil is used the query needs to only pull data with host.name (which I think is just packetbeat) otherwise the model is going to be somewhat useless for any network data that isn't provided by packetbeat (e.g. netflow, vpc flow, firewwall logs, sensors, etc.)  

Either the partitioning needs to be removed or set to source.ip, or the detection needs to be specific to data that includes host.name.  Otherwise the partitioning will be effectively based on packet beat data by host.name, and everything else (including non packetbeat data for the same hosts if any other network data is being ingested. 

from an enterprise standpoint the idea of deploying packetbeat / network capture wideliy, instead of sensors or netflow, is resource prohibitive, and nothing I would ever want to implement in the real world.

dainperkins commented:
Also if Packetbeat is used as a sensor (e.g. spanning a network tap) every doc will have the same host.name (the host name of the device receiving the network span)

sodhikirti07 commented:
> Either the partitioning needs to be removed or set to source.ip, or the detection needs to be specific to data that includes host.name.

@dainperkins the partitionings on network exfil jobs are already removed. 

> USB exfil - from a security and behavior modeling standpoint, in any environment with shared workstations this will work much better if partitioned by user, as hosts will invariably be less aligned to a specific behavior profile, and it seemed like most file events will have a user involved

From an ML standpoint, we think it's better to use `host.name` as a partitioning field for USB based jobs. Also, the influencers section adds the user related info to the alerts and anomalies.

dainperkins commented:
sorry - thought the following was in regards to **all** the jobs:

> Based on my discussion with @susan-shu-c and @ajosh0504, I am keeping the current partition_by configurations in place i.e., partitioning the jobs by host.name.
